concordium-smart-contract-testing/src/invocation/impls.rs

```diff
@@ -0,0 +1,1468 @@
+    /// provided `invoked_contract` was changed.
+    ///
+    /// **Preconditions:**
+    ///  - All contracts, modules, accounts referred must exist in persistence.
+    ///  - All amount deltas must be valid (i.e. not cause underflows when added
+    ///    to balance).
+    pub(crate) fn persist(
+        mut self,
+        remaining_energy: Energy,
+        invoked_contract: ContractAddress,
+        persisted_accounts: &mut BTreeMap<AccountAddress, Account>,
+        persisted_contracts: &mut BTreeMap<ContractAddress, Contract>,
+    ) -> Result<(Energy, bool), OutOfEnergy> {
+        let current = self.current_mut();
+        let mut invoked_contract_has_state_changes = false;
+        // Persist contract changes and collect the total increase in states sizes.
+        let mut collector = v1::trie::SizeCollector::default();
+        let mut loader = v1::trie::Loader::new(&[][..]);
+
+        let mut frozen_states: BTreeMap<ContractAddress, trie::PersistentState> = BTreeMap::new();
+
+        // Create frozen versions of all the states, to compute the energy needed.
+        for (addr, changes) in current.contracts.iter_mut() {
+            if let Some(modified_state) = &mut changes.state {
+                frozen_states.insert(*addr, modified_state.freeze(&mut loader, &mut collector));
+            }
+        }
+
+        // One energy per extra byte of state.
+        let energy_for_state_increase = Energy::from(collector.collect());
+
+        // Return an error if out of energy, and clear the changeset.
+        if remaining_energy
+            .checked_sub(energy_for_state_increase)
+            .is_none()
+        {
+            return Err(OutOfEnergy);
+        }
+
+        // Then persist all the changes.
+        for (addr, changes) in current.contracts.iter_mut() {
+            let mut contract = persisted_contracts
+                .get_mut(addr)
+                .expect("Precondition violation: contract must exist");
+            // Update balance.
+            if !changes.self_balance_delta.is_zero() {
+                contract.self_balance = changes
+                    .self_balance_delta
+                    .apply_to_balance(changes.self_balance_original)
+                    .expect("Precondition violation: amount delta causes underflow");
+            }
+            // Update module reference.
+            if let Some(new_module_ref) = changes.module {
+                contract.module_reference = new_module_ref;
+            }
+            // Update state.
+            if changes.state.is_some() {
+                if *addr == invoked_contract {
+                    invoked_contract_has_state_changes = true;
+                }
+                // Replace with the frozen state we created earlier.
+                contract.state = frozen_states
+                    .remove(addr)
+                    .expect("Known to exist since we just added it.");
```

Comment:
In principle we don't have to panic here. We just `;` and ignore the issue.

Reply:
`contract.state` is of type `PersistentState` not `Option<PersistentState>`, so we do need some kind of unwrapping.